### PR TITLE
@ #164 | should update nodejs dev guide with support for nodejs debugging

### PR DESCRIPTION
**Pull request by hoatle**
connect #164 

- need to review technical content (@hieptranquoc)
- need to review English content (@hoavt)

### CODE REVIEW

**Review by phuonglm — DISMISSED** [review]

File: docs/nodejs_dev_guide.rst [hunk 1]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
-Add Docker files
-----------------
+Create dev mode
+---------------
 
-We need to add:
-  - ``Dockerfile-dev`` and ``docker-compose.yml`` files for dev mode
-  - ``Dockerfile`` and ``docker-compose.prod.yml`` files for prod mode
+Dev mode usually contains development packages to assist development productivity, for example:
+auto reload when code changes; debugging, etc.
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: there are code changes, debugging, etc.

File: docs/nodejs_dev_guide.rst [hunk 2]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+
+
+The changes should be like this:
+https://github.com/acme101/nodejs-hello-world/commit/b0a7ac4e95898ecd12651830917a5f4db9561420
+
+
+Run on dev mode
+---------------
+
+Open a new terminal window, ``ssh`` into the ``teracy-dev`` VM to execute the following commands:
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: should use ``vagrant ssh`` for clearer

File: docs/nodejs_dev_guide.rst [hunk 3]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+
+After that, open:
+
+- http://dev.nodejs.teracy.dev or https://dev.nodejs.teracy.dev on your host browser to see the app
+  on the dev mode.
+- Check out the VM's :ref:`basic_usage-ip_address` and on any device within your LAN,
+  open http://dev.nodejs.<vm_ip>.xip.io or https://dev.nodejs.<vm_ip>.xip.io to see the web app.
+
+
+2 way sync for node_modules
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: two-way sync for node_modules

File: docs/nodejs_dev_guide.rst [hunk 4]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+---------------------------
+
+By default, we disable ``node_modules`` sync. To have 2 way sync, we need to configure it.
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: To have 2-way sync,

File: docs/nodejs_dev_guide.rst [hunk 5]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+
+- Create ``node_modules`` directory from the host terminal window:
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: Create the ``node_modules`` directory

File: docs/nodejs_dev_guide.rst [hunk 6]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+
+  ..  code-block:: bash
+
+      $ cd ~/teracy-dev/workspace/nodejs-hello-world
+      $ mkdir node_modules
+      $ touch node_modules/.gitkeep
+
+- Add gitignore for node_modules content, except .gitkeep to keep the empty directory, this is required
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: Add ``gitignore`` for the ``node_modules`` content, except ``.gitkeep`` to keep the empty directory. This is required
by the configuration below.

Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: The configuration below is "vagrant_config_override.json"?

File: docs/nodejs_dev_guide.rst [hunk 7]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+  by the configuration below.
+
+- Configure vagrant_config_override.json the following content:
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: Configure the ``vagrant_config_override.json`` file with the following content:

File: docs/nodejs_dev_guide.rst [hunk 8]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+          "_id": "2",
+          "options": {
+            "aliases": ["dev.nodejs.teracy.dev", "review.nodejs.teracy.dev" "nodejs.teracy.dev"]
+          }
+        }]
+      }
+
+- ``$ vagrant reload`` to get it take effect.
+
+- Re-run the dev container to get it install npm packages again:
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: to get it to install the ``npm`` packages again:

File: docs/nodejs_dev_guide.rst [hunk 9]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+
+..  code-block:: bash
+
+    $ vagrant ssh
+    $ ws
+    $ cd nodejs-hello-world
+    $ docker-compose restart dev && docker-compose logs -f dev
+
+Afer this, npm packages should be syned into the host ``node_modules`` directory.
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: ``npm``

File: docs/nodejs_dev_guide.rst [hunk 10]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+When you scale the dev services into more containers, you can do the same by attaching more debug hosts into the IDE.
+
+
+Create Prod mode
+----------------
+
+Prod mode will run the Docker image of the app which is used for production deployment. The Docker image
+usually contains only the run-time stuff.
+
+- Create ``Dockerfile`` file within the ``nodejs-hello-world`` directory with the following content:
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: Create the...

File: docs/nodejs_dev_guide.rst [hunk 11]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+          yarn cache clean
 
-      CMD npm run start
+      ADD . $APP
 
-- Create ``docker-compose.prod.yml`` file within the ``hello-world-nodejs`` directory with the
+      CMD ["sh", "run-prod.sh"]
+
+
+- Create ``docker-compose.prod.yml`` file within the ``nodejs-hello-world`` directory with the
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: Create the...

File: docs/nodejs_dev_guide.rst [hunk 12]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+https://github.com/acme101/nodejs-hello-world/commit/a710fda1e9602e4f5e558198c6a3206affb8976e
 
-Run on dev mode
----------------
 
-Open a new terminal window, `ssh` into the ``teracy-dev`` VM to execute the following commands:
+Run on prod mode
+----------------
+
+Open a new terminal window,, `ssh` into the ``teracy-dev`` VM to execute the following commands:
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: should use ``vagrant ssh`` for be clearer

File: docs/nodejs_dev_guide.rst [hunk 13]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+      docker-compose -f docker-compose.yml -f docker-compose.prod.yml logs -f prod # to run the production Docker image
 
 
-After that, open hello-d.teracy.dev on your browser to see the app on the dev mode.
+After that, open:
 
-Run on prod mode
-----------------
+- http://nodejs.teracy.dev or https://nodejs.teracy.dev on your host browser to see the app
+  on the dev mode.
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: dev mode or prod mode?

File: docs/nodejs_dev_guide.rst [hunk 14]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+
+The changes should be like this: https://github.com/acme101/nodejs-hello-world/commit/1f7d0cfdd82435a482834fb45b2203674134cfc2
+
+
+Run on review mode
+------------------
+
+For example, we're going to review `hoatle/nodejs-hello-world:feature-1` Docker image.
 
 Open a new terminal window,, `ssh` into the ``teracy-dev`` VM to execute the following commands:
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: ``vagrant ssh``

File: docs/nodejs_dev_guide.rst [hunk 15]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
+-------------------
+
+It's required that we should always run CI/CD to automate the build and deployment. In this guide,
+we're going to create CI/CD for travis-ci and gitlab-ci.
+
+For deployment, we're going to deploy on Heroku, Google Container Engine (Kubernertes).
+
+The changes should be like this: https://github.com/acme101/nodejs-hello-world/commit/ca822a679691de619200e7cd2a0a5d946e5045ae
+
+More the most up to date and more information, please checkout the README.md file from the
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: To have the most up-to-date  and more information, please checkout the ``README.md`` file....

File: docs/nodejs_dev_guide.rst [hunk 16]
```diff
@@ -215,148 +199,359 @@ Add app.js and update package.json's scripts
...
 
 
-After that, open hello.teracy.dev on your browser to see the app on the prod mode.
+Summary
+-------
 
-Congratulations, you've created a basic hello world Node.js app with Docker workflow running
-on the ``teracy-dev``.
+Congratulations, we've created a basic hello world Node.js app with Docker workflow, CI/CD system on
+the ``teracy-dev``. This is the current best practices to work with ``teracy-dev``, we can apply
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: This is --> These are

**Review by hoavt — COMMENTED** [review]
@hoatle: Please check my comments

**Comment by hoatle** [comment]
@phuonglm don't approve (blindly?) like that, you should approve only after @hieptranquoc and @hoavt's reviewing.

**Comment by hoatle** [comment]
@hoavt feedbacks applied

**Comment by hoatle** [comment]
@hieptranquoc need your check on the technical content is there is any problem.

File: docs/nodejs_dev_guide.rst [hunk 17]
```diff
@@ -541,7 +541,7 @@ For deployment, we're going to deploy on Heroku, Google Container Engine (Kubern
 
 The changes should be like this: https://github.com/acme101/nodejs-hello-world/commit/ca822a679691de619200e7cd2a0a5d946e5045ae
 
-More the most up to date and more information, please checkout the README.md file from the
+More the most up-to-date and more information, please checkout the README.md file from the
```
Comment (hoavt) [inline docs/nodejs_dev_guide.rst]: --> To get the most....

**Review by hoavt — COMMENTED** [review]
@hoatle: Please check my comments

**Comment by hieptranquoc** [comment]
@hoatle reviewed, technical content is ok

**Comment by hoavt** [comment]
It's ok to merge.

**Comment by hoatle** [comment]
@hoavt please approve.

**Review by phuonglm — APPROVED** [review]

**Review by hoavt — APPROVED** [review]

**Pull request merged**